Sources/AsyncHTTPClient/ConnectionPool/HTTP2/HTTP2Connection.swift

```diff
@@ -258,16 +266,25 @@ final class HTTP2Connection {
     private func shutdown0() {
         self.channel.eventLoop.assertInEventLoop()
 
-        self.state = .closing
+        switch self.state {
+        case .active, .starting:
+            self.state = .closing
 
-        // inform all open streams, that the currently running request should be cancelled.
-        self.openStreams.forEach { box in
-            box.channel.triggerUserOutboundEvent(HTTPConnectionEvent.shutdownRequested, promise: nil)
-        }
+            // inform all open streams, that the currently running request should be cancelled.
+            self.openStreams.forEach { box in
+                box.channel.triggerUserOutboundEvent(HTTPConnectionEvent.shutdownRequested, promise: nil)
+            }
 
-        // inform the idle connection handler, that connection should be closed, once all streams
-        // are closed.
-        self.channel.triggerUserOutboundEvent(HTTPConnectionEvent.shutdownRequested, promise: nil)
+            // inform the idle connection handler, that connection should be closed, once all streams
+            // are closed.
+            self.channel.triggerUserOutboundEvent(HTTPConnectionEvent.shutdownRequested, promise: nil)
+            
+        case .closed:
+            // we are already closed and we need to tolerate this
+            break
+        case .initialized, .closing:
+            preconditionFailure("invalid state \(self.state)")
```

Comment:
I think we should also support shutdown calls if we are already closing. No reason to crash here. Also do we want to support shutdown, if we haven't even connected? We could go straight to `closed`, which is the only way to drop a connection currently.

Reply:
Agree, changed in https://github.com/swift-server/async-http-client/pull/646/commits/653858d01b989f9f206d29ebf9e87f0898dd01b3